core/src/com/fallenflame/game/TextController.java

```diff
@@ -0,0 +1,87 @@
+package com.fallenflame.game;
+
+import com.badlogic.gdx.Gdx;
+import com.badlogic.gdx.graphics.Color;
+import com.badlogic.gdx.graphics.g2d.BitmapFont;
+import com.badlogic.gdx.graphics.g2d.GlyphLayout;
+import com.badlogic.gdx.math.Rectangle;
+import com.badlogic.gdx.math.Vector2;
+import com.badlogic.gdx.utils.JsonValue;
+import com.fallenflame.game.util.JsonAssetManager;
+import org.w3c.dom.css.Rect;
+
+import java.util.*;
+
+public class TextController {
+    private Map<Rectangle, String> texts;
+    private String nextMessage = null;
+    private BitmapFont displayFont;
+    private int screenWidth;
+    private int screenHeight;
+
+    public void initialize(JsonValue jsonValue) {
+        if (jsonValue == null) return;
+        texts = new HashMap<>();
+        for (JsonValue textJson : jsonValue) {
+            float[] i = textJson.get("rect").asFloatArray();
+            Rectangle r = new Rectangle(i[0] - i[2] / 2 , i[1] - i[3] / 2, i[2], i[3]);
+            texts.put(r, textJson.get("text").asString());
+        }
+        displayFont = JsonAssetManager.getInstance().getEntry("display", BitmapFont.class);
+        screenWidth = Gdx.graphics.getWidth();
+        screenHeight = Gdx.graphics.getHeight();
+    }
+
+    public void dispose() {
+        texts.clear();
+        texts = null;
+    }
+
+    public void update(PlayerModel player) {
+        if (texts == null) return;
+        for (Map.Entry<Rectangle, String> ele : texts.entrySet()) {
+            if (ele.getKey().contains(player.getX(), player.getY())) {
+                nextMessage = ele.getValue();
+                return;
+            }
+        }
```

Comment:
So if I'm reading this right, do the player X and Y coordinates have to be exactly the value stored in order to transition the message? Having it required to be an exact match may be challenging for the player to actually trigger

Reply:
It's not an exact value. The key is a rectangle so it is a box.